File: src/gestorAplicacion/usuario/Coordinador.java
package gestorAplicacion.usuario;

import java.util.ArrayList;
import gestorAplicacion.administracion.*;
import java.io.Serializable;
// import gestorAplicacion.usuario.*;

public class Coordinador extends Usuario implements Serializable{
    private String facultad;
    private final static int limitesCreditos=20;
    // private static ArrayList<Materia> materiasTotales;
    private static final long serialVersionUID = 1L;

    public Coordinador(String facultad,long id, String nombre, String pw) {
        super(id, nombre,pw);
        super.setTipo("Coordinador");
        this.facultad = facultad;
    }


    //Metodos

    //Abstract

    public String toString(){
        return "Nombre: "+ getNombre()+ "\nDocumento: "+ getId();
    }

    // Ya se encuentra hecha mas abajo :P
    // public void crearMateria(String nombre, int codigo, int creditos){
    //     // Generarlo aqui o en el main?
    //     Materia materia = new Materia(nombre,codigo, "Sin descripcion", creditos, "Sin facultad", 0);
    //     // materiasTotales.add(materia);
    // }

    public String desmatricular(Estudiante estudiante, Grupo grupo){
        // Desmatricular no seria simplemente llamar un metodo de estudiante?
        // Tambien seria que dentro de materias de estudiante no este esta materia

        boolean estaMatriculado = grupo.existenciaEstudiante(estudiante);

        if (estaMatriculado){
            grupo.eliminarEstudiante(estudiante);
            return "El estudiante ha sido desmatriculado de la materia y su respectivo grupo";
        }
        else{
            return "El estudiante no estaba matriculado";
        }
    }

    public void contratarProfesor(String nombre, String facultad){
        Profesor profesor = new Profesor(nombre,facultad);
    }


    // FUSIONE resturarEstudiantesInscritosEnMateria Y resturarProfesoresInvolucradosEnMateria EN resturarMateria.
    public void resturarMateria(Materia materia){
        for (int i=0;i<materia.getGrupos().size();i++){
            Grupo puntero_Grupo = materia.getGrupos().get(i);
            puntero_Grupo.getProfesor().desvincularGrupo(puntero_Grupo);

            for (int j=0;j<puntero_Grupo.getEstudiantes().size();j++){
                Estudiante puntero_Estudiante = puntero_Grupo.getEstudiantes().get(j);
                this.desmatricular(puntero_Estudiante, puntero_Grupo);
                // puntero_Estudiante.desmatricularMateria(materia);
            }   
        }
    }


    // ligadura dinamica?
    public static Object[] crearHorarioAleatorio(ArrayList<Materia> materias){
        /**
         * Toma una lista de materias que se desean ver.
         * 
         * Crear un horario aleatorio en base de los grupos disponibles.
         * 
         * Retorna una lista estatica de dos elementos: Un booleano que nos dira si fue posible o no
         *  crear el horaio, el horario generado y la materia que no permitio crear el horario en caso de existir.
         */

        Object[] resultado = new Object[3];
                
        Horario horario =  new Horario();
        boolean ok = true;
        Materia materiaObstaculo = null;


        int[] gPosible = new int[materias.size()];        
        int i =0; // indice de materias
        
        while (true){
            ArrayList<String> pClases = materias.get(i).getGrupos().get(gPosible[i]).getHorario();
            if (!horario.comprobarDisponibilidad(pClases)){
                if(gPosible[i]==materias.get(i).getGrupos().size()-1){
                    ok = false;
                    horario = null;
                    materiaObstaculo = materias.get(i);
                    break;
                }
                else{
                    gPosible[i]++;
                }
            }
            
            else if(i==materias.size()-1){
                for (int k=0;k<materias.size();k++){
                    Grupo grupo = materias.get(k).getGrupos().get(gPosible[k]);
                    horario.ocuparHorario(grupo);
                }
                break;
            }
            
            else{
                i++;
            };
            // }
        }

        resultado[0] = ok;
        resultado[1] = horario;
        resultado[2] = materiaObstaculo;

        return resultado;

    }

    /*Metodo eliminarMateria: Recibira una materia y rectificara see encuentre en la base de datos quiere eliminar de la base de datos
    sí se encuentre en esta, y además eliminarla correctamente de las materias de los estudiantes 
    que la tienen inscrita y de los profesores relacionados a esta*/

    public void eliminarMateria(Materia materia){
        if(Materia.getMateriasTotales().contains(materia)){
            Materia.getMateriasTotales().remove(materia);
            resturarMateria(materia);
        }
        
    }

    /*Metodo agregarMateria: Recibira los parámetros necesarios para crear una materia, si esta no se encuentra en
    la base de datos, la creara con sus respectivos atributos*/
    public void agregarMateria(String nombre, int codigo, String descripcion,int creditos, String facultad, int cupos, ArrayList<Materia> prerrequisitos, ArrayList<Grupo> grupos){
        for (Materia materia : Materia.getMateriasTotales()){
            if (materia.getNombre().equals(nombre) == false){
                Materia nMateria = new Materia(nombre, codigo, descripcion, creditos, facultad, cupos, prerrequisitos, grupos);
            }
        }
    }
    /*Metodo candidato a beca: ... */
    public boolean candidatoABeca(Estudiante estudiante, Beca tipoDeBeca){
        if (tipoDeBeca.getCupos() > 0){
            if ((estudiante.calcularPromedio >= tipoDeBeca.getPromedioRequerido()) && (estudiante.calcularAvance >= tipoDeBeca.getAvanceRequerido()) && (estudiante.getCreditos() >= tipoDeBeca.getCreditosInscritosRequeridos())){
                if (tipoDeBeca.necesitaRecomendacion){
                    if (recomendarEstudiante(estudiante)){
                        return true;
                    }else return false;
                } else return true; //no necesita recomendacion, pero cumple los demás requisitos
            }
            else return false;
        }
        else return false;
    }
        

                                                                                                                     
    // Getters - Setters

    public String getFacultad() {
        return this.facultad;
    }
        
    public static int getLimitesCreditos() {
        return limitesCreditos;
    }

    public void setFacultad(String facultad) {
        this.facultad = facultad;
    }
}

File: Python/src/gestorAplicacion/usuario/Coordinador.py
from gestorAplicacion.administracion.Horario import Horario
from gestorAplicacion.administracion.Materia import Materia
from gestorAplicacion.usuario.Estudiante import Estudiante
from gestorAplicacion.usuario.Profesor import Profesor
from gestorAplicacion.usuario.Usuario import Usuario
from gestorAplicacion.administracion.Beca import Beca
from excepciones.ErrorManejo import *
from excepciones.ObjetoInexistente import *

class Coordinador(Usuario):
    _usuarioIngresado = None
    _LIMITES_CREDITOS = 20
    _coordinadoresTotales = []
    _facultades = [
        "Facultad de arquitectura",
        "Facultad de ciencias",
        "Facultad de ciencias agrarias",
        "Facultad de ciencias humanas y economicas",
        "Facultad de minas",
        "Sede",
    ]

    def __init__(self, facultad, id, nombre, pw):
        super().__init__(id, nombre, facultad, pw)
        super().setTipo("Coordinador")
        Coordinador._coordinadoresTotales.append(self)

    # METODOS

    @staticmethod
    def desmatricular(estudiante, grupo):
        estaMatriculado = grupo.existenciaEstudiante(estudiante)

        if estaMatriculado:
            grupo.eliminarEstudiante(estudiante)
            return "El estudiante ha sido desmatriculado de la materia y su respectivo grupo"
        else:
            return "El estudiante no estaba matriculado"

    @staticmethod
    def restaurarMateria(materia):
        for i in range(len(materia.getGrupos())):
            puntero_Grupo = materia.getGrupos()[i]
            puntero_Grupo.getProfesor().desvincularGrupo(puntero_Grupo)

            for j in range(len(puntero_Grupo.getEstudiantes())):
                puntero_Estudiante = puntero_Grupo.getEstudiantes()[j]
                Coordinador.desmatricular(puntero_Estudiante, puntero_Grupo)


    def desmatricularDelSistema(self, estudiante):
        e1 = None
        for e in Estudiante.getEstudiantes():
            if e == estudiante:
                e1 = e

        if e1 is not None:
            Estudiante.getEstudiantes().remove(e1)

        for usuario in Usuario.getUsuariosTotales():
            if isinstance(usuario, Estudiante):
                if usuario == estudiante:
                    Usuario.getUsuariosTotales().remove(usuario)
                    break

    def crearHorario(materias):
        """
        Toma una lista de materias que se desean ver.

        Crea un horario aleatorio basado en los grupos disponibles.

        Retorna una lista estática de dos elementos: Un booleano que nos dirá si fue posible o no
        crear el horario, el horario generado y la materia que no permitió crear el horario en caso de existir.
        """

        resultado = [None, None, None]
        horario = Horario()
        ok = True
        materiaObstaculo = None

        gPosible = [0] * len(materias)
        mPosibles = [0] * len(materias)
        i = 0  # índice de materias

        while True:
            pClases = materias[i].getGrupos()[gPosible[i]].getHorario()
            if horario.comprobarDisponibilidad(pClases):
                # Asignamos la materia al horario
                horario.ocuparHorario(materias[i].getGrupos()[gPosible[i]])
                # Al ponerse 1 significa que sí es posible poner dicha materia
                mPosibles[i] = 1
                # Pasamos a la siguiente materia
                i += 1
                # Comprobamos si es la última
                if i == len(materias):
                    break
            else:
                # Pasamos al siguiente grupo en la materia i
                gPosible[i] += 1

                # Comprobamos si la materia es imposible de dar
                if gPosible[i] == len(materias[i].getGrupos()):
                    # Tenemos que probar todas las posibilidades, por lo tanto probamos con el siguiente grupo de la materia i-1
                    i -= 1
                    horario.liberarHorario(
                        materias[i].getGrupos()[gPosible[i]].getHorario()
                    )
                    gPosible[i] += 1
                    gPosible[i + 1] = 0

                    # Comprobamos si, aunque iteramos todas las posibilidades, no se puede poner la materia i
                    if gPosible[i] == len(materias[i].getGrupos()):
                        m = 0
                        for k in mPosibles:
                            if k == 0:
                                materiaObstaculo = materias[m]
                                ok = False
                            else:
                                m += 1
                        break

        resultado[0] = ok
        resultado[1] = horario
        resultado[2] = materiaObstaculo

        return resultado

    
    def eliminarMateria(self, materia):
        if materia in Materia.getMateriasTotales():
            Coordinador.restaurarMateria(materia)
            Materia.getMateriasTotales().remove(materia)
        else:
            raise CampoVacio("Se ha ingresado un valor inválido")

    
    def agregarMateria(self, nombre, codigo, descripcion, creditos, facultad, prerrequisitos):
        nombreMaterias = []

        for materia in Materia.getMateriasTotales():
            nombreMaterias.append(materia.getNombre())

        if nombre not in nombreMaterias:
            nMateria = Materia(
                nombre, codigo, descripcion, creditos, facultad, prerrequisitos
            )

    @classmethod
    def candidatoABeca(cls,estudiante, tipoDeBeca):
        if tipoDeBeca.getCupos() > 0:
            if (
                estudiante.getPromedio() >= tipoDeBeca.getPromedioRequerido()
                and estudiante.getAvance() >= tipoDeBeca.getAvanceRequerido()
                and estudiante.getCreditos()
                >= tipoDeBeca.getCreditosInscritosRequeridos()
                and estudiante.getEstrato() <= tipoDeBeca.getEstratoMinimo()
            ):
                if tipoDeBeca.getNecesitaRecomendacion():
                    if Profesor.recomendarEstudiante(estudiante):
                        return True
                    else:
                        return False
                else:
                    return True  # No necesita recomendación, pero cumple los demás requisitos
            else:
                return False
        else:
            return False
    @classmethod
    def mostrarFacultades(cls):
        retorno = ""
        i = 1
        for facultad in Coordinador._facultades:
            retorno += str(i) + ". " + facultad + "\n"
            i += 1
        return retorno

    def __str__(self):
        return "Nombre Coordinador: " + self.getNombre() + "\nDocumento: " + self.getId()

    # Setters y Getters

    @classmethod
    def getLimitesCreditos(cls):
        return cls._LIMITES_CREDITOS

    @classmethod
    def getCoordinadoresTotales(cls):
        return cls._coordinadoresTotales

    @classmethod
    def setCoordinadoresTotales(cls, coordinadores):
        cls._coordinadoresTotales = coordinadores

    @classmethod
    def getFacultades(cls):
        return cls._facultades

    @classmethod
    def setFacultades(cls, facultades):
        cls._facultades = facultades
    
    @classmethod
    def mostrarBecas(cls):
        i = 1
        for beca in Beca.getBecas():
            a = beca.getConvenio()
    
    @classmethod
    def getUsuarioIngresado(cls):
        return cls._usuarioIngresado
    
    @classmethod
    def setUsuarioIngresado(cls, usuario):
        cls._usuarioIngresado = usuario